### DISCUSSION TITLE
API Documentation

### DISCUSSION THREAD

**Discussion by pdifeo** [rustdesk/rustdesk#5834] [Q&A]
is the list of APIs and their documentation available ?

The only thing I have found is the possibility of assigning on the command line with the parameters --assign and --token the user associated with the device. 

I would like to do more.

**Comment by rustdesk** [comment]
We do not need to hide it if it is available. :)

**Reply by pdifeo** [reply]
The implied question is: will it be ?

**Reply by rustdesk** [reply]
Currently, we are thinking more about how to rewrite our documentation totally, our documentation is very bad from my standard and users' feedback.

**Reply by rustdesk** [reply]
I knew your implied question, my implied answer is no ETA yet. :)

**Reply by pdifeo** [reply]
You will have realised that mine was intended as a friendly provocation :-) 

I would like to contribute to the project in my own small way by looking at the monitoring side for security purposes.

For example, reading logs in order to be able to send alarms about possible problems with Zabbix, which I use to my satisfaction in other situations.

Before using RustDesk on my customers' computers, I have to make sure that everything is working properly. Also because I have to convince them to abandon TeamViewer (I have had a licence for 10 years) for RustDesk.

Since I don't have the source code of the pro version, I can't reverse engineer it, so I ask.

Regards
Pasquale

**Reply by rustdesk** [reply]
Let me give you something for reverse engineer. Almost all apis are ready in web console, but with different naming. You can inspect like this. As below, replace `peers` with `devices` in your api call. Additionly, we use `guid` in api, but `id` is used somewhere in the web console.

![image](https://github.com/rustdesk/rustdesk/assets/71636191/3ae1d751-c477-42d9-87a1-674a5238fa04)


## <web_console_host>/static/index.html#/groups
```
         .route("/api/groups", get(groups))
         .route("/api/groups/:guid", delete(delete_group_by_guid))
         .route("/api/groups", post(new_group))
         .route("/api/groups/:guid", patch(update_group_by_guid))
```

## <web_console_host>/static/index.html#/audit/<type>
```
         .route("/api/audits/conn", get(audits_conn))
         .route("/api/audits/file", get(audits_file))
         .route("/api/audits/alarm", get(audits_alarm))
         .route("/api/audits/console", get(audits_console))
```

## <web_console_host>/static/index.html#/devices
```
         .route("/api/devices", get(peers))
         .route("/api/devices/cli", post(update_peer_cli)) // call from RustDesk client from command line with api token
         .route("/api/devices/:guid", patch(update_peer_by_guid))
         .route("/api/devices/:guid", delete(delete_peer_by_guid))
```

## <web_console_host>/static/index.html#/users
```
         .route("/api/users", get(users))
         .route("/api/users/invite", post(invite_user))
         .route("/api/users/:guid", patch(update_user_by_guid))
         .route("/api/users/:guid", delete(delete_user))
         .route("/api/users", post(new_user))
```

## <web_console_host>/static/index.html#/strategy
```
         .route("/api/strategies", get(strategies))
         .route("/api/strategies/:guid/assign", put(assign_strategy_by_guid))
         .route("/api/strategies/assign", post(assign_strategy))
```

**Reply by rustdesk** [reply]
I like parsec's api documentation. https://parsec.app/docs/teams-api

**Reply by pdifeo** [reply]
Good.

I did a test.
1) I generated a read-only token
2) I tried the call to
`https://<my-server-address>/api/groups`
with  
**POST** method 
**Bearer authorisation** with the created token
**Content-Type** header to **application/json**
3) I received a reply 
```
{
    "total: 5
}
```
which is the number of groups that are created on my server
4) I added to the call
`current=1&pageSize=20`
and as a response I got
```
{
    "data": [
        {
            "guid": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "name": "GRP01",
            "team": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "created_at": 1695754674
        },
        {
            "guid": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "name": "GRP02",
            "team": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "created_at": 1693824983
        },
        {
            "guid": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "name": "GRP03",
            "team": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "created_at": 1693673547
        },
        {
            "guid": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "name": "GRP04",
            "team": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "created_at": 1693672829
        },
        {
            "guid": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "name": "Default",
            "team": "xxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxx",
            "created_at": 1693646025
        }
    ],
    "total": 5
}
```
This weekend or later I will try to create documentation with Swagger.

Greetings
Pasquale

**Reply by BertBrenner** [reply]
Is there any way to list addressbooks and peers within addressboks with a token?

**Reply by rustdesk** [reply]
Not yet, we will add it later and update here.

**Reply by yaroz** [reply]
> Is there any way to list addressbooks and peers within addressboks with a token?

If you use the method outlined above, and make a few changes, you can see what the body and uri should be.  For instance, I just built a powershell function that gets the addressbook entries, cycles through each one, getting the device_name from the device listing (matching IDs), then getting user information from AD based on the device name, then updating the Alias and tags.  The URI should be 'http://<relay>:21114/api/ab/peer/update/<addressbookuri>'.  You'll also have to get your token from the request headers (in chrome), as there is no token for address book manipulation right now.

**Reply by CryptoZyphix** [reply]
> > Is there any way to list addressbooks and peers within addressboks with a token?
> 
> If you use the method outlined above, and make a few changes, you can see what the body and uri should be. For instance, I just built a powershell function that gets the addressbook entries, cycles through each one, getting the device_name from the device listing (matching IDs), then getting user information from AD based on the device name, then updating the Alias and tags. The URI should be 'http://:21114/api/ab/peer/update/'. You'll also have to get your token from the request headers (in chrome), as there is no token for address book manipulation right now.

yaroz do you mind sharing your PS script? I'm having difficulty getting my batch script to add a device to an address book based on their documentation here https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/console/ You can also assign address book in this way, e.g. --assign --token <generatedtoken> --address_book_name <addressbookname> or --assign --token <generatedtoken> --address_book_name <addressbookname> --address_book_tag <addressbooktag>

**Reply by stewba11** [reply]
Hi @rustdesk ,

Has there been any progress on adding in Token functionality for updating the address book? We manage over 200 sites and so far I have been able to automate everything, except updating the password on the device in the Address Book. I have built the required scripts to do it, I just need a Token for auth. 

Just as an overview, I have a Powershell script inside an Intune deployment that will install RustDesk, add the device to the appropriate shared address book and generate a random local password. The last step is updating the device password in the shared address book itself.

Coming from TeamViewer we have been able to get everything we need (and more!) from RustDesk, if we can just get this one last feature added in, we will have a fantastic replacement!

Thank you for building an awesome product!

**Comment by CryptoZyphix** [comment]
@yaroz do you mind sharing your PS script? I'm having difficulty getting my batch script to add a device to an address book based on their documentation here https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/console/ You can also assign address book in this way, e.g. --assign --token --address_book_name or --assign --token --address_book_name --address_book_tag

**Reply by yaroz** [reply]
@CryptoZyphix Sorry.. I saw the request 2 weeks ago and forgot to do it.  Here's what I was using before we just used custom packages.  Stupid markdown... https://pastebin.com/nQg7ajj7

**Reply by OtterKring** [reply]
Nice! Thx!

We are having a PoC of RustDesk atm and I converted the users.py and devices.py to Powershell, in case someone can use it. If we get out of PoC and go live (IF), I will make a real module if I find time.
[rustdesk-powershell.zip](https://github.com/user-attachments/files/19285616/rustdesk-powershell.zip)

I successfully tested view, enable and disable up to now. Delete did not work, even after running disable first. No idea why, yet.

Did I get this right from your script, that using method "Patch" will rewrite all device data connected to an ID?
